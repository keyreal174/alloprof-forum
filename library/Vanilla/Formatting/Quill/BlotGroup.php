<?php
/**
 * @author Adam (charrondev) Charron <adam.c@vanillaforums.com>
 * @copyright 2009-2019 Vanilla Forums Inc.
 * @license GPL-2.0-only
 */

namespace Vanilla\Formatting\Quill;

use Vanilla\Formatting\Quill\Blots\AbstractBlot;
use Vanilla\Formatting\Quill\Blots\Embeds\ExternalBlot;
use Vanilla\Formatting\Quill\Blots\Lines\AbstractLineTerminatorBlot;
use Vanilla\Formatting\Quill\Blots\CodeLineBlot;
use Vanilla\Formatting\Quill\Blots\Lines\CodeLineTerminatorBlot;
use Vanilla\Formatting\Quill\Blots\Lines\HeadingTerminatorBlot;
use Vanilla\Formatting\Quill\Blots\Lines\ParagraphLineTerminatorBlot;
use Vanilla\Formatting\Quill\Blots\TextBlot;

/**
 * Class to represent a group of a quill blots. One group can contain:
 *
 * - Multiple Text formats in a single paragraph (as well as inline embeds).
 * - Multiple line blots if it is a line type grouping
 */
class BlotGroup {
    /** @var AbstractBlot[] */
    private $blots = [];

    /**
     * @var AbstractBlot[]
     *
     * Blots that can determine the surrounding tag over the other blot types.
     */
    private $overridingBlots = [
        CodeLineTerminatorBlot::class,
        AbstractLineTerminatorBlot::class,
    ];

    /**
     * Determine if this group is made up only of a break.
     *
     * @return bool
     */
    public function isBreakOnlyGroup(): bool {
        if (count($this->blots) !== 1) {
            return false;
        }

        $blot = $this->blots[0];

        return get_class($blot) === TextBlot::class && $this->blots[0]->getContent() === "";
    }

    /**
     * Check if the group's last matches the passed Blot class.
     *
     * @param string $blotClass - The class constructor of a Blot.
     * @param bool $needsExactMatch - If true, the class must match exactly, no child classes are allowed.
     *
     * @return bool
     */
    public function endsWithBlotOfType(string $blotClass, bool $needsExactMatch = false): bool {
        if (count($this->blots) === 0) {
            return false;
        }

        $lastBlot = $this->blots[count($this->blots) - 1];

        if ($needsExactMatch) {
            return get_class($lastBlot) === $blotClass;
        }

        return $lastBlot instanceof $blotClass;
    }

    /**
     * Determine if the group is empty.
     *
     * @return bool
     */
    public function isEmpty(): bool {
        return count($this->blots) === 0;
    }


    /**
     * Render the block.
     *
     * @return string
     */
    public function render(): string {
        // We need a special exception for empty paragraph only blots.
        // This because the ParagraphLineTerminatorBlot does not render a break if there is other group content
        // But does render a break if it is alone. This is mostly to make styling easier.
        // We do not want to need make the paragraph terminator aware of its group.
        if (
            count($this->blots) === 1 &&
            $this->blots[0] instanceof ParagraphLineTerminatorBlot &&
            $this->blots[0]->getContent() === "\n"
        ) {
            return "<p><br></p>";
        }

        // Don't render empty groups.
        $surroundTagBlot = $this->getBlotForSurroundingTags();
        $result = $surroundTagBlot->getGroupOpeningTag();

        // Line blots have special rendering.
        if ($surroundTagBlot instanceof AbstractLineTerminatorBlot) {
            $result .= $this->renderLineGroup();
        } else {
            foreach ($this->blots as $blot) {
                $result .= $blot->render();
            }
        }

        $result .= $surroundTagBlot->getGroupClosingTag();
        return $result;
    }

    /**
     * Render out a group with line terminators.
     *
     * We need to render potentially multiple inline blots than a line terminator, but the line terminator also
     * is responsible for the opening tag of the line and each line can have its special attributes even if
     * they are the same blot (eg. list indentation).
     *
     * @return string
     */
    public function renderLineGroup(): string {
        $result = "";

        // Look ahead and get all of the line terminators.
        $terminatorIndex = 0;
        $terminators = $this->getLineTerminators();
        $terminator = $terminators[$terminatorIndex];

        // Grab the first line terminator and start the line.
        $result .= $terminator->renderLineStart();

        foreach ($this->blots as $index => $blot) {
            if ($blot instanceof AbstractLineTerminatorBlot) {
                // Render out the content of the line terminator (maybe nothing, maybe extra newlines).
                $result .= $terminator->render();
                // End the line.
                $result .= $terminator->renderLineEnd();

                // Start the next line.
                $terminatorIndex++;
                $terminator = $terminators[$terminatorIndex] ?? null;
                if ($terminator !== null) {
                    $result .= $terminator->renderLineStart();
                }
            } else {
                // Render out inline blots.
                $result .= $blot->render();
            }
        }

        return $result;
    }

    /**
     * Get all of the line terminators in the group.
     *
     * @return array
     */
    private function getLineTerminators(): array {
        $terminators = [];
        foreach ($this->blots as $blot) {
            if ($blot instanceof AbstractLineTerminatorBlot) {
                $terminators[] = $blot;
            }
        }
        return $terminators;
    }

    /**
     * Get all of the mention blots in the group.
     *
     * Mentions that are inside of Blockquote's are excluded. We don't want to be sending notifications when big quote
     * replies build up.
     *
     * @return string[]
     */
    public function getMentionUsernames() {
        if ($this->getBlotForSurroundingTags() instanceof Blots\Lines\BlockquoteLineTerminatorBlot) {
            return [];
        }

        $names = [];
        foreach ($this->blots as $blot) {
            if ($blot instanceof Blots\Embeds\MentionBlot) {
                $names[] = $blot->getUsername();
            }
        }

        return $names;
    }

    /**
     * Add a blot to this block.
     *
     * @param AbstractBlot[] $blots
     */
    public function pushBlots(array $blots) {
        foreach ($blots as $blot) {
            $this->blots[] = $blot;
        }
    }

    /**
     * Get the position in the blots array of the first blot of the given type. Defaults to -1.
     *
     * @param string $blotClass The class string of the blot to check for.
     *
     * @return int
     */
    public function getIndexForBlotOfType(string $blotClass): int {
        $index = -1;

        foreach ($this->blots as $blotIndex => $blot) {
            if ($blot instanceof $blotClass) {
                $index = $blotIndex;
                break;
            }
        }

        return $index;
    }

    /**
     * Determine which blot should create the surrounding HTML tags of the group.
     *
     * @return AbstractBlot|null
     */
    public function getBlotForSurroundingTags() {
        if (count($this->blots) === 0) {
            return null;
        }
        $blot = $this->blots[0];
        $overridingBlot = $this->getPrimaryBlot();

        return $overridingBlot ?? $blot;
    }

    /**
     * Get the primary blot for the group.
     *
     * @return null|AbstractBlot
     */
    public function getPrimaryBlot() {
        foreach ($this->overridingBlots as $overridingBlot) {
            $index = $this->getIndexForBlotOfType($overridingBlot);
            if ($index >= 0) {
                return $this->blots[$index];
            }
        }

        return null;
    }

    /**
     * Simplify the blot group into an array of the classnames of its blots.
     *
     * @return string[]
     */
    public function getTestData(): array {
        $blots = [];
        foreach ($this->blots as $blot) {
            $blots[] = [
                "class" => get_class($blot),
                "content" => $blot->getContent(),
            ];
        }

        return $blots;
    }

    /**
     * Get the text value of a heading. This is not sanitized, so be sure to HTML escape before using.
     *
     * @return string
     */
    public function getUnsafeText(): string {
        $text = "";
        foreach ($this->blots as $blot) {
            if ($blot instanceof TextBlot
                || $blot instanceof ExternalBlot) {
                $text .= $blot->getContent();
            }
        }

        return $text;
    }
}
