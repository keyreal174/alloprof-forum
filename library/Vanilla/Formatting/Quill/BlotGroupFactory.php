<?php
/**
 * @author Adam (charrondev) Charron <adam.c@vanillaforums.com>
 * @copyright 2009-2018 Vanilla Forums Inc.
 * @license https://opensource.org/licenses/GPL-2.0 GPL-2.0
 */

namespace Vanilla\Formatting\Quill;

use Vanilla\Formatting\Quill\Blots\AbstractBlot;
use Vanilla\Formatting\Quill\Blots\Lines\AbstractLineBlot;

class BlotGroupFactory {

    /** @var array[] */
    private $operations;
    /** @var string[]  */
    private $allowedBlotClasses;
    /** @var string */
    private $parseMode = Parser::PARSE_MODE_NORMAL;
    /** @var AbstractBlot */
    private $inProgressBlot;
    /** @var BlotGroup */
    private $inProgressGroup;
    /** @var array[] */
    private $inProgressLine;
    /** @var BlotGroup[] */
    private $groups;

    private $currentOp;
    private $nextOp;
    private $prevOp;

    /**
     * @return BlotGroup[]
     */
    public function getGroups(): array {
        return $this->groups;
    }

    /**
     * BlotGroupFactory constructor.
     *
     * @param array[] $operations
     * @param string[] $allowedBlotClasses
     */
    public function __construct(array $operations, array $allowedBlotClasses) {
        $this->operations = $operations;
        $this->allowedBlotClasses = $allowedBlotClasses;
        $this->inProgressGroup = new BlotGroup();
        $this->inProgressLine = [];
        $this->groups = [];
        $this->createBlotGroups();
    }

    private function clearLine() {
        $this->inProgressGroup->pushBlots($this->inProgressLine);
        $this->inProgressLine = [];
    }

    private function clearBlotGroup() {
        if (!$this->inProgressGroup->isEmpty()) {
            $this->groups[] = $this->inProgressGroup;
            $this->inProgressGroup = new BlotGroup();
        }
    }


    /**
     * Create blot groups out of an array of operations.
     *
     * @param array $operations The pre-parsed operations
     * @param string $parseMode The parse mode to create blots with.
     *
     * @return BlotGroup[]
     */
    private function createBlotGroups() {
        $operationLength = count($this->operations);

        for ($i = 0; $i < $operationLength; $i++) {
            $this->currentOp = $this->operations[$i];
            $this->prevOp = $this->operations[$i - 1] ?? [];
            $this->nextOp = $this->operations[$i + 1] ?? [];
            $this->inProgressBlot = $this->getCurrentBlot();


            // In event of break blots we want to clear the group if applicable and the skip to the next item.
            if ($this->currentOp === Parser::BREAK_OPERATION) {
                $this->clearLine();
                $this->clearBlotGroup();
                continue;
            }

            // Ask the blot if it should close the current group.
            if ($this->inProgressBlot instanceof AbstractLineBlot) {
                if (($this->inProgressBlot->shouldClearCurrentGroup($this->inProgressGroup))) {
                    $this->clearBlotGroup();
                }
                $this->clearLine();
            }

            $this->inProgressLine[] = $this->inProgressBlot;

            if ($this->inProgressBlot instanceof AbstractLineBlot) {
                $this->clearLine();
            }

            // Some block type blots get a group all to themselves.
            if ( $this->inProgressBlot->isOwnGroup()) {
                $this->clearLine();
                $this->clearBlotGroup();
            }
        }
        $this->clearLine();
        if (!$this->inProgressGroup->isEmpty()) {
            $this->groups[] = $this->inProgressGroup;
        }
    }

    /**
     * Get the matching blot for a sequence of operations. Returns the default if no match is found.

     * @return AbstractBlot
     */
    public function getCurrentBlot(): AbstractBlot {
        // Fallback to a TextBlot if possible. Otherwise we fallback to rendering nothing at all.
        $blotClass = Blots\TextBlot::matches($this->currentOp) ? Blots\TextBlot::class : Blots\NullBlot::class;
        foreach ($this->allowedBlotClasses as $blot) {
            // Find the matching blot type for the current, last, and next operation.
            if ($blot::matches($this->currentOp)) {
                $blotClass = $blot;
                break;
            }
        }

        return new $blotClass($this->currentOp, $this->prevOp, $this->nextOp, $this->parseMode);
    }
}
