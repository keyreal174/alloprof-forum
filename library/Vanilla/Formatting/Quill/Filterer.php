<?php
/**
 * @author Adam Charron <adam.c@vanillaforums.com>
 * @copyright 2009-2019 Vanilla Forums Inc.
 * @license GPL-2.0-only
 */

namespace Vanilla\Formatting\Quill;

use Vanilla\EmbeddedContent\Embeds\QuoteEmbed;
use Vanilla\Formatting\Exception\FormattingException;

/**
 * Class for filtering Rich content before it gets inserted into the database.
 */
class Filterer {

    /**
     * Filter the contents of a quill post.
     *
     * - Validates content.
     * - Strips useless embed data.
     *
     * @param string $content The content to filter.
     * @return string
     *
     * @throws FormattingException If the JSON could not be converted to operations.
     */
    public function filter(string $content): string {
        $operations = Parser::jsonToOperations($content);
        // Re-encode the value to escape unicode values.
        $operations = $this->cleanupEmbeds($operations);
        $operations = json_encode($operations);
        return $operations;
    }


    /**
     * There is certain embed data from the rich editor that we want to strip out. This includes
     *
     * - Malformed partially formed operations (dataPromise).
     * - Nested embed data.
     *
     * @param array[] $operations The quill operations to loop through.
     * @return array
     */
    private function cleanupEmbeds(array &$operations): array {
        foreach ($operations as $key => &$op) {
            if (!is_array($op['insert']['embed-external'] ?? null)) {
                continue;
            }
            $embed = &$op['insert']['embed-external'];

            // If a dataPromise is still stored on the embed, that means it never loaded properly on the client.
            // We want to strip these embeds that haven't finished properly loading.
            $dataPromise = $embed['dataPromise'] ?? null;
            if ($dataPromise !== null) {
                unset($operations[$key]);
            }

            if (!is_array($embed['data'] ?? null)) {
                // We only care about embeds operations.
                continue;
            }
            $embedData = &$embed['data'];


            if (!$embedData) {
                // Clean up that messed up operation.
                $operations[$key];
                continue;
            }

            // Remove the rendered bodies. The raw bodies are the source of truth.
            $format = &$embedData['format'] ?? null;
            $bodyRaw = &$embedData['bodyRaw'] ?? null;
            $type = &$embedData['type'] ?? null;

            if ($type !== QuoteEmbed::TYPE) {
                // We only care about quote embeds specifically.
                continue;
            }

            // Remove nested external embed data. We don't want it rendered and this will prevent it from being
            // searched.
            if ($format === 'Rich' && is_array($bodyRaw)) {
                // Iterate through the nested embed.
                foreach ($bodyRaw as $subInsertIndex => &$subInsertOp) {
                    $insert = &$subInsertOp['insert'];
                    if (is_array($insert)) {
                        $url = $insert['embed-external']['data']['url'] ?? null;
                        if ($url !== null) {
                            // Replace the embed with just a link.
                            $linkEmbedOps = $this->makeLinkEmbedInserts($url);
                            array_splice($bodyRaw, $subInsertIndex, 1, $linkEmbedOps);
                        }
                    }
                }
            }

            // Finally render the new body to overwrite the previous HTML body.
            // We also need to ensure we've safely rendered the body to prevent innacurate content.
            $embedData['body'] = \Gdn_Format::quoteEmbed($bodyRaw, $format);
        }

        return array_values($operations);
    }

    /**
     * Make the contents of a link embed.
     *
     * @param string $url
     * @return array
     */
    private function makeLinkEmbedInserts(string $url): array {
        return [
            [
                'insert' => $url,
                'attributes' => [
                    'link' => $url,
                ],
            ],
            [ 'insert' => "\n" ],
        ];
    }
}
