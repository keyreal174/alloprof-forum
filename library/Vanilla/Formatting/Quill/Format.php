<?php
/**
 * @author Adam Charron <adam.c@vanillaforums.com>
 * @copyright 2009-2019 Vanilla Forums Inc.
 * @license GPL-2.0-only
 */

namespace Vanilla\Formatting\Quill;

use Garden\StaticCacheTranslationTrait;
use Vanilla\Formatting\BaseFormat;
use Vanilla\Formatting\Exception\FormattingException;
use Vanilla\Formatting\Heading;
use Vanilla\Formatting\Quill\Blots\Lines\HeadingTerminatorBlot;
use Vanilla\Web\Html\TwigRenderTrait;

class Format extends BaseFormat {

    use TwigRenderTrait;
    use StaticCacheTranslationTrait;

    /** @var string */
    const RENDER_ERROR_MESSAGE = 'There was an error rendering this rich post.';

    /** @var Parser */
    private $parser;

    /** @var Renderer */
    private $renderer;

    /** @var Filterer */
    private $filterer;

    /**
     * Constructor for DI.
     *
     * @param Parser $parser
     * @param Renderer $renderer
     * @param Filterer $filterer
     */
    public function __construct(Parser $parser, Renderer $renderer, Filterer $filterer) {
        $this->parser = $parser;
        $this->renderer = $renderer;
        $this->filterer = $filterer;
    }


    /**
     * @inheritdoc
     */
    public function renderHTML(string $content): string {
        try {
            $operations = Parser::jsonToOperations($content);
        } catch (FormattingException $e) {
            return $this->renderErrorMessage();
        }

        $blotGroups = $this->parser->parse($operations);
        return $this->renderer->render($blotGroups);
    }

    /**
     * @inheritdoc
     */
    public function renderPlainText(string $content): string {
        $text = '';
        try {
            $operations = Parser::jsonToOperations($content);
        } catch (FormattingException $e) {
            return self::t(self::RENDER_ERROR_MESSAGE);
        }

        $blotGroups = $this->parser->parse($operations);

        /** @var BlotGroup $blotGroup */
        foreach ($blotGroups as $blotGroup) {
            $text .= $blotGroup->getUnsafeText();
        }
        return $text;
    }

    /**
     * @inheritdoc
     */
    public function filter(string $content): string {
        return $this->filterer->filter($content);
    }

    /**
     * @inheritdoc
     */
    public function parseAttachments(string $content): array {
        return [];
    }

    /**
     * @inheritdoc
     */
    public function parseHeadings(string $content): array {
        $outline = [];

        try {
            $operations = Parser::jsonToOperations($content);
        } catch (FormattingException $e) {
            return [];
        }

        $parser = (new Parser())
            ->addBlot(HeadingTerminatorBlot::class);
        $blotGroups = $parser->parse($operations);

        /** @var BlotGroup $blotGroup */
        foreach ($blotGroups as $blotGroup) {
            $blot = $blotGroup->getPrimaryBlot();
            if ($blot instanceof HeadingTerminatorBlot && $blot->getReference()) {
                $outline[] = new Heading(
                    $blotGroup->getUnsafeText(),
                    $blot->getHeadingLevel(),
                    $blot->getReference()
                );
            }
        }
        return $outline;
    }

    /**
     * Render an error message indicating something went wrong.
     *
     * @return string
     */
    private function renderErrorMessage(): string {
        $data = [
            'title' => self::t(self::RENDER_ERROR_MESSAGE),
            'errorUrl' => 'https://docs.vanillaforums.com/help/addons/rich-editor/#why-is-my-published-post-replaced-with-there-was-an-error-rendering-this-rich-post',
        ];

        return $this->renderTwig('resources/userContentError', $data);
    }
}
