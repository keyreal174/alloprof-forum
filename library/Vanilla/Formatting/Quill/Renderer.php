<?php
/**
 * @author Adam (charrondev) Charron <adam.c@vanillaforums.com>
 * @copyright 2009-2018 Vanilla Forums Inc.
 * @license https://opensource.org/licenses/GPL-2.0 GPL-2.0
 */

namespace Vanilla\Formatting\Quill;

/**
 * Class for rendering BlotGroups into HTML.
 */
class Renderer {

    /**
     * Render operations into HTML.
     *
     * @param BlotGroup[] $blotGroups The blots to render.
     *
     * @return string
     */
    public function render(array $blotGroups): string {
        $result = "";
        foreach ($blotGroups as $index => $group) {
            $result .= $group->render();
        }

        return $result;
    }

    /**
     * Render operations into HTML.
     *
     * @param BlotGroup[] $blotGroups The blots to render.
     *
     * @return string
     */
    public function renderQuote(array $blotGroups): string {
        $result = "";
        foreach ($blotGroups as $index => $group) {
            $result .= $group->renderQuote();
        }

        return $result;
    }
}
