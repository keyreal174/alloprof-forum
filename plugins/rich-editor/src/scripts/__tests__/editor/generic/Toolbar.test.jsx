/**
 * @author Adam (charrondev) Charron <adam.c@vanillaforums.com>
 * @copyright 2009-2018 Vanilla Forums Inc.
 * @license https://opensource.org/licenses/GPL-2.0 GPL-2.0
 */

import React from "react";
import { mount, shallow } from "enzyme";
import Quill from "quill/core";
import EditorMenuItem from "../../../editor/generic/MenuItem";
import { Toolbar } from "../../../editor/generic/Toolbar";

jest.mock('quill/core');

test("matches snapshot", () => {
    const quill = new Quill();
    const toolbar = shallow(
        <Toolbar quill={quill}/>
    );

    expect(toolbar).toMatchSnapshot();
});

test("generates correct number of <EditorMenuItem /> Components", () => {
    const menuItems = {
        foo: {
            active: false,
        },
        bar: {
            active: false,
        },
        other: {
            active: true,
        },
        thing: {
            active: false,
        },
    };
    const quill = new Quill();
    const toolbar = shallow(
        <Toolbar quill={quill} menuItems={menuItems}/>
    );

    expect(toolbar.find(EditorMenuItem).length).toBe(4);
});

test("can receive a custom formatter for the menu item click handler.", () => {
    const mockFormatter = jest.fn();

    const menuItems = {
        bold: {
            active: false,
            formatter: mockFormatter,
        },
    };
    const quill = new Quill();
    const toolbar = mount(
        <Toolbar quill={quill} menuItems={menuItems}/>
    );

    toolbar.find(".richEditor-button").simulate("click");
    toolbar.find(".richEditor-button").simulate("click");

    expect(mockFormatter.mock.calls.length).toBe(2);
});

describe("update", () => {
    let toolbar;

    beforeAll(() => {
        const quill = new Quill();
        const menuItems = {
            bold: {
                active: false,
            },
        };

        toolbar = shallow(
            <Toolbar quill={quill} menuItems={menuItems}/>
        );
    });

    it("updates boolean type ", () => {

    });
});
