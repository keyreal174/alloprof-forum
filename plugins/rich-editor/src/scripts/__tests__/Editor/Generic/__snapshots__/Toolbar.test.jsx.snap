// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Toolbar
    quill={
      Quill {
        "addContainer": [MockFunction],
        "blur": [MockFunction],
        "deleteText": [MockFunction],
        "disable": [MockFunction],
        "enable": [MockFunction],
        "focus": [MockFunction],
        "format": [MockFunction],
        "formatLine": [MockFunction],
        "formatText": [MockFunction],
        "getBounds": [MockFunction],
        "getContents": [MockFunction],
        "getFormat": [MockFunction],
        "getIndex": [MockFunction],
        "getLeaf": [MockFunction],
        "getLength": [MockFunction],
        "getLine": [MockFunction],
        "getLines": [MockFunction],
        "getModule": [MockFunction],
        "getSelection": [MockFunction],
        "getText": [MockFunction],
        "hasFocus": [MockFunction],
        "insertEmbed": [MockFunction],
        "insertText": [MockFunction],
        "isEnabled": [MockFunction],
        "off": [MockFunction],
        "on": [MockFunction] {
          "calls": Array [
            Array [
              "editor-change",
              [Function],
            ],
            Array [
              "scroll-optimize",
              [Function],
            ],
          ],
        },
        "once": [MockFunction],
        "pasteHTML": [MockFunction],
        "removeFormat": [MockFunction],
        "scrollIntoView": [MockFunction],
        "setContents": [MockFunction],
        "setSelection": [MockFunction],
        "setText": [MockFunction],
        "update": [MockFunction],
        "updateContents": [MockFunction],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="richEditor-menuItems MenuItems"
      >
        <MenuItem
          isActive={false}
          isFirst={true}
          isLast={false}
          label="richEditor.menu.bold"
          onBlur={undefined}
          onClick={[Function]}
          propertyName="bold"
          role={undefined}
        />
        <MenuItem
          isActive={false}
          isFirst={false}
          isLast={false}
          label="richEditor.menu.italic"
          onBlur={undefined}
          onClick={[Function]}
          propertyName="italic"
          role={undefined}
        />
        <MenuItem
          isActive={false}
          isFirst={false}
          isLast={false}
          label="richEditor.menu.strike"
          onBlur={undefined}
          onClick={[Function]}
          propertyName="strike"
          role={undefined}
        />
        <MenuItem
          isActive={false}
          isFirst={false}
          isLast={false}
          label="richEditor.menu.code"
          onBlur={undefined}
          onClick={[Function]}
          propertyName="code"
          role={undefined}
        />
        <MenuItem
          isActive={false}
          isFirst={false}
          isLast={true}
          label="richEditor.menu.link"
          onBlur={undefined}
          onClick={[Function]}
          propertyName="link"
          role={undefined}
        />
      </div>,
      "className": "richEditor-menu",
      "role": "menu",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <MenuItem
            isActive={false}
            isFirst={true}
            isLast={false}
            label="richEditor.menu.bold"
            onBlur={undefined}
            onClick={[Function]}
            propertyName="bold"
            role={undefined}
          />,
          <MenuItem
            isActive={false}
            isFirst={false}
            isLast={false}
            label="richEditor.menu.italic"
            onBlur={undefined}
            onClick={[Function]}
            propertyName="italic"
            role={undefined}
          />,
          <MenuItem
            isActive={false}
            isFirst={false}
            isLast={false}
            label="richEditor.menu.strike"
            onBlur={undefined}
            onClick={[Function]}
            propertyName="strike"
            role={undefined}
          />,
          <MenuItem
            isActive={false}
            isFirst={false}
            isLast={false}
            label="richEditor.menu.code"
            onBlur={undefined}
            onClick={[Function]}
            propertyName="code"
            role={undefined}
          />,
          <MenuItem
            isActive={false}
            isFirst={false}
            isLast={true}
            label="richEditor.menu.link"
            onBlur={undefined}
            onClick={[Function]}
            propertyName="link"
            role={undefined}
          />,
        ],
        "className": "richEditor-menuItems MenuItems",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "class",
          "props": Object {
            "isActive": false,
            "isFirst": true,
            "isLast": false,
            "label": "richEditor.menu.bold",
            "onBlur": undefined,
            "onClick": [Function],
            "propertyName": "bold",
            "role": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "class",
          "props": Object {
            "isActive": false,
            "isFirst": false,
            "isLast": false,
            "label": "richEditor.menu.italic",
            "onBlur": undefined,
            "onClick": [Function],
            "propertyName": "italic",
            "role": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "class",
          "props": Object {
            "isActive": false,
            "isFirst": false,
            "isLast": false,
            "label": "richEditor.menu.strike",
            "onBlur": undefined,
            "onClick": [Function],
            "propertyName": "strike",
            "role": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "3",
          "nodeType": "class",
          "props": Object {
            "isActive": false,
            "isFirst": false,
            "isLast": false,
            "label": "richEditor.menu.code",
            "onBlur": undefined,
            "onClick": [Function],
            "propertyName": "code",
            "role": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "4",
          "nodeType": "class",
          "props": Object {
            "isActive": false,
            "isFirst": false,
            "isLast": true,
            "label": "richEditor.menu.link",
            "onBlur": undefined,
            "onClick": [Function],
            "propertyName": "link",
            "role": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="richEditor-menuItems MenuItems"
        >
          <MenuItem
            isActive={false}
            isFirst={true}
            isLast={false}
            label="richEditor.menu.bold"
            onBlur={undefined}
            onClick={[Function]}
            propertyName="bold"
            role={undefined}
          />
          <MenuItem
            isActive={false}
            isFirst={false}
            isLast={false}
            label="richEditor.menu.italic"
            onBlur={undefined}
            onClick={[Function]}
            propertyName="italic"
            role={undefined}
          />
          <MenuItem
            isActive={false}
            isFirst={false}
            isLast={false}
            label="richEditor.menu.strike"
            onBlur={undefined}
            onClick={[Function]}
            propertyName="strike"
            role={undefined}
          />
          <MenuItem
            isActive={false}
            isFirst={false}
            isLast={false}
            label="richEditor.menu.code"
            onBlur={undefined}
            onClick={[Function]}
            propertyName="code"
            role={undefined}
          />
          <MenuItem
            isActive={false}
            isFirst={false}
            isLast={true}
            label="richEditor.menu.link"
            onBlur={undefined}
            onClick={[Function]}
            propertyName="link"
            role={undefined}
          />
        </div>,
        "className": "richEditor-menu",
        "role": "menu",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <MenuItem
              isActive={false}
              isFirst={true}
              isLast={false}
              label="richEditor.menu.bold"
              onBlur={undefined}
              onClick={[Function]}
              propertyName="bold"
              role={undefined}
            />,
            <MenuItem
              isActive={false}
              isFirst={false}
              isLast={false}
              label="richEditor.menu.italic"
              onBlur={undefined}
              onClick={[Function]}
              propertyName="italic"
              role={undefined}
            />,
            <MenuItem
              isActive={false}
              isFirst={false}
              isLast={false}
              label="richEditor.menu.strike"
              onBlur={undefined}
              onClick={[Function]}
              propertyName="strike"
              role={undefined}
            />,
            <MenuItem
              isActive={false}
              isFirst={false}
              isLast={false}
              label="richEditor.menu.code"
              onBlur={undefined}
              onClick={[Function]}
              propertyName="code"
              role={undefined}
            />,
            <MenuItem
              isActive={false}
              isFirst={false}
              isLast={true}
              label="richEditor.menu.link"
              onBlur={undefined}
              onClick={[Function]}
              propertyName="link"
              role={undefined}
            />,
          ],
          "className": "richEditor-menuItems MenuItems",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "class",
            "props": Object {
              "isActive": false,
              "isFirst": true,
              "isLast": false,
              "label": "richEditor.menu.bold",
              "onBlur": undefined,
              "onClick": [Function],
              "propertyName": "bold",
              "role": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "class",
            "props": Object {
              "isActive": false,
              "isFirst": false,
              "isLast": false,
              "label": "richEditor.menu.italic",
              "onBlur": undefined,
              "onClick": [Function],
              "propertyName": "italic",
              "role": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "2",
            "nodeType": "class",
            "props": Object {
              "isActive": false,
              "isFirst": false,
              "isLast": false,
              "label": "richEditor.menu.strike",
              "onBlur": undefined,
              "onClick": [Function],
              "propertyName": "strike",
              "role": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "3",
            "nodeType": "class",
            "props": Object {
              "isActive": false,
              "isFirst": false,
              "isLast": false,
              "label": "richEditor.menu.code",
              "onBlur": undefined,
              "onClick": [Function],
              "propertyName": "code",
              "role": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "4",
            "nodeType": "class",
            "props": Object {
              "isActive": false,
              "isFirst": false,
              "isLast": true,
              "label": "richEditor.menu.link",
              "onBlur": undefined,
              "onClick": [Function],
              "propertyName": "link",
              "role": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
